File: packages/sdk/src/events.ts
import type * as TrueForge from "./api/index.js";

/**
 * Returns true when `event` is a streaming delta that must be merged into a
 * base `model.message` with the same `id`.
 */
export function isEventDelta(event: { type: string }): event is TrueForge.ModelMessageDeltaEvent {
    return event.type === "model.message.delta";
}

/**
 * Merges a `model.message.delta` into its base `model.message` in place.
 *
 * Deltas and their base share an `id`. Content / reasoning text is appended;
 * tool-call fragments are merged by `index`; `finishReason`, `usage`, and
 * `refusal` overwrite when present on the delta.
 *
 * No-ops when `base` is not a `model.message` or the ids do not match.
 */
export function mergeEventDelta(
    base: TrueForge.TurnStreamingEvent | TrueForge.ModelMessageEvent,
    delta: TrueForge.ModelMessageDeltaEvent,
): void {
    if (base.type !== "model.message") {
        return;
    }
    if (base.id !== delta.id) {
        return;
    }

    if (delta.content != null && delta.content.length > 0) {
        if (base.content == null || typeof base.content === "string") {
            base.content = `${base.content ?? ""}${delta.content}`;
        }
    }

    if (delta.reasoningContent != null && delta.reasoningContent.length > 0) {
        base.reasoningContent = `${base.reasoningContent ?? ""}${delta.reasoningContent}`;
    }

    if (delta.refusal !== undefined) {
        base.refusal = delta.refusal;
    }

    if (delta.finishReason !== undefined) {
        base.finishReason = delta.finishReason;
    }

    if (delta.usage !== undefined) {
        base.usage = delta.usage;
    }

    if (delta.toolCalls != null && delta.toolCalls.length > 0) {
        if (base.toolCalls == null) {
            base.toolCalls = [];
        }
        for (const deltaToolCall of delta.toolCalls) {
            mergeToolCallDelta(base.toolCalls, deltaToolCall);
        }
    }
}

function mergeToolCallDelta(
    toolCalls: TrueForge.ToolCall[],
    deltaToolCall: TrueForge.ExtendedChunkDeltaToolCall,
): void {
    const { index } = deltaToolCall;
    const existing = toolCalls[index];

    if (existing !== undefined) {
        if (deltaToolCall.id !== undefined) {
            existing.id = deltaToolCall.id;
        }
        if (deltaToolCall.toolInfo !== undefined) {
            existing.toolInfo = deltaToolCall.toolInfo;
        }
        if (deltaToolCall.providerSpecificFields !== undefined) {
            existing.providerSpecificFields = deltaToolCall.providerSpecificFields;
        }
        if (deltaToolCall.function !== undefined) {
            if (deltaToolCall.function.name !== undefined) {
                existing.function.name = deltaToolCall.function.name;
            }
            if (deltaToolCall.function.arguments !== undefined) {
                existing.function.arguments += deltaToolCall.function.arguments;
            }
        }
        return;
    }

    // Only append sequentially — sparse / out-of-order indexes are ignored until
    // earlier slots exist (streaming tool calls are emitted in index order).
    if (index !== toolCalls.length) {
        return;
    }

    const id = deltaToolCall.id;
    const name = deltaToolCall.function?.name;
    const toolInfo = deltaToolCall.toolInfo;
    if (id === undefined || name === undefined || toolInfo === undefined) {
        return;
    }

    const created: TrueForge.ToolCall = {
        id,
        type: "function",
        function: {
            name,
            arguments: deltaToolCall.function?.arguments ?? "",
        },
        toolInfo,
    };
    if (deltaToolCall.providerSpecificFields !== undefined) {
        created.providerSpecificFields = deltaToolCall.providerSpecificFields;
    }
    toolCalls.push(created);
}

File: python/trueforge_sdk/src/trueforge_sdk/events.py
from __future__ import annotations

import typing

from .types.chat_completion_message_tool_call_function import ChatCompletionMessageToolCallFunction
from .types.extended_chunk_delta_tool_call import ExtendedChunkDeltaToolCall
from .types.model_message_delta_event import ModelMessageDeltaEvent
from .types.model_message_event import ModelMessageEvent
from .types.tool_call import ToolCall
from .types.turn_streaming_event import TurnStreamingEvent


def is_event_delta(event: object) -> typing.TypeGuard[ModelMessageDeltaEvent]:
    return getattr(event, "type", None) == "model.message.delta"


def merge_event_delta(
    base: TurnStreamingEvent | ModelMessageEvent,
    delta: ModelMessageDeltaEvent,
) -> None:
    """Merge a `model.message.delta` into its matching `model.message` in place."""
    if base.type != "model.message":
        return
    if base.id != delta.id:
        return

    if delta.content is not None and len(delta.content) > 0:
        if base.content is None or isinstance(base.content, str):
            base.content = f"{base.content or ''}{delta.content}"

    if delta.reasoning_content is not None and len(delta.reasoning_content) > 0:
        base.reasoning_content = f"{base.reasoning_content or ''}{delta.reasoning_content}"

    if delta.refusal is not None:
        base.refusal = delta.refusal

    if delta.finish_reason is not None:
        base.finish_reason = delta.finish_reason

    if delta.usage is not None:
        base.usage = delta.usage

    if delta.tool_calls is not None and len(delta.tool_calls) > 0:
        if base.tool_calls is None:
            base.tool_calls = []
        for delta_tool_call in delta.tool_calls:
            _merge_tool_call_delta(base.tool_calls, delta_tool_call)


def _merge_tool_call_delta(
    tool_calls: list[ToolCall],
    delta_tool_call: ExtendedChunkDeltaToolCall,
) -> None:
    index = delta_tool_call.index
    if index < len(tool_calls):
        existing = tool_calls[index]
        if delta_tool_call.id is not None:
            existing.id = delta_tool_call.id
        if delta_tool_call.tool_info is not None:
            existing.tool_info = delta_tool_call.tool_info
        if delta_tool_call.provider_specific_fields is not None:
            existing.provider_specific_fields = delta_tool_call.provider_specific_fields
        if delta_tool_call.function is not None:
            if delta_tool_call.function.name is not None:
                existing.function.name = delta_tool_call.function.name
            if delta_tool_call.function.arguments is not None:
                existing.function.arguments += delta_tool_call.function.arguments
        return

    # Streaming tool calls are emitted in index order; ignore sparse gaps.
    if index != len(tool_calls):
        return

    tool_id = delta_tool_call.id
    name = None if delta_tool_call.function is None else delta_tool_call.function.name
    tool_info = delta_tool_call.tool_info
    if tool_id is None or name is None or tool_info is None:
        return

    arguments = ""
    if delta_tool_call.function is not None and delta_tool_call.function.arguments is not None:
        arguments = delta_tool_call.function.arguments

    created = ToolCall(
        id=tool_id,
        type="function",
        function=ChatCompletionMessageToolCallFunction(name=name, arguments=arguments),
        tool_info=tool_info,
    )
    if delta_tool_call.provider_specific_fields is not None:
        created.provider_specific_fields = delta_tool_call.provider_specific_fields
    tool_calls.append(created)
